dlt/destinations/impl/clickhouse/clickhouse.py

```diff
@@ -0,0 +1,402 @@
+        else:
+            raise LoadJobTerminalException(
+                file_path,
+                f"ClickHouse loader does not support '{bucket_scheme}' filesystem.",
+            )
+
+        statement = f"INSERT INTO {qualified_table_name} SELECT * FROM {table_function}"
+        with client.begin_transaction():
+            client.execute_sql(statement)
+
+    def state(self) -> TLoadJobState:
+        return "completed"
+
+    def exception(self) -> str:
+        raise NotImplementedError()
+
+
+class ClickHouseMergeJob(SqlMergeJob):
+    @classmethod
+    def _to_temp_table(cls, select_sql: str, temp_table_name: str) -> str:
+        return f"CREATE TABLE {temp_table_name} ENGINE = Memory AS {select_sql};"
+
+    @classmethod
+    def gen_key_table_clauses(
+        cls,
+        root_table_name: str,
+        staging_root_table_name: str,
+        key_clauses: Sequence[str],
+        for_delete: bool,
+    ) -> List[str]:
+        join_conditions = " AND ".join([c.format(d="d", s="s") for c in key_clauses])
+        return [
+            f"FROM {root_table_name} AS d JOIN {staging_root_table_name} AS s ON {join_conditions}"
+        ]
+
+    @classmethod
+    def gen_update_table_prefix(cls, table_name: str) -> str:
+        return f"ALTER TABLE {table_name} UPDATE"
+
+    @classmethod
+    def requires_temp_table_for_delete(cls) -> bool:
+        return False
+
+
+class ClickHouseClient(SqlJobClientWithStaging, SupportsStagingDestination):
+    capabilities: ClassVar[DestinationCapabilitiesContext] = capabilities()
+
+    def __init__(
+        self,
+        schema: Schema,
+        config: ClickHouseClientConfiguration,
+    ) -> None:
+        self.sql_client: ClickHouseSqlClient = ClickHouseSqlClient(
+            config.normalize_dataset_name(schema), config.credentials
+        )
+        super().__init__(schema, config, self.sql_client)
+        self.config: ClickHouseClientConfiguration = config
+        self.active_hints = deepcopy(HINT_TO_CLICKHOUSE_ATTR)
+        self.type_mapper = ClickHouseTypeMapper(self.capabilities)
+
+    def _create_merge_followup_jobs(self, table_chain: Sequence[TTableSchema]) -> List[NewLoadJob]:
+        return [ClickHouseMergeJob.from_table_chain(table_chain, self.sql_client)]
+
+    def _get_column_def_sql(self, c: TColumnSchema, table_format: TTableFormat = None) -> str:
+        # Build column definition.
+        # The primary key and sort order definition is defined outside column specification.
+        hints_str = " ".join(
+            self.active_hints.get(hint)
+            for hint in self.active_hints.keys()
+            if c.get(hint, False) is True
+            and hint not in ("primary_key", "sort")
+            and hint in self.active_hints
+        )
+
+        # Alter table statements only accept `Nullable` modifiers.
+        # JSON type isn't nullable in ClickHouse.
+        type_with_nullability_modifier = (
```

Comment:
Clickhouse will not fail when updating a table to add a non-nullable column to a table. I think we should probably prevent update statements that contain non-nullable columns in _make_add_column_sql for all destinations so make the behavior uniform. There is one test that is not passing because of this (adding a dlt_root_it to an existing table does not raise an exception on clickhouse at this point)

Reply:
dremio has the same behavior and I think we are skipping this check in the tests. my take would be to generate log warning that such thing happened (ALTER with NOT NULL) and let destination handle it. if the table is empty adding NOT NULL will work. also some weird things like automatic default value and whatever may be the case. So I'd let destination handle this

Reply:
we may add another capability

Reply:
the logger already prints a warning in the superclass if you do this.